pkg/envoy/options.go

```diff
@@ -0,0 +1,758 @@
+		}
+		return nil
+	},
+})
+
+// ServiceCluster sets the --service-cluster flag, which defines the local service cluster
+// name where Envoy is running
+func ServiceCluster(c string) Option {
+	return &genericOption{
+		value: c,
+		v:     serviceClusterValidator,
+	}
+}
+
+var serviceClusterValidator = registerFlagValidator(&flagValidator{
+	flagName: "--service-cluster",
+})
+
+// ServiceNode sets the --service-node flag, which defines the local service node name
+// where Envoy is running
+func ServiceNode(n string) Option {
+	return &genericOption{
+		value: n,
+		v:     serviceNodeValidator,
+	}
+}
+
+var serviceNodeValidator = registerFlagValidator(&flagValidator{
+	flagName: "--service-node",
+})
+
+// DrainDuration sets the --drain-time-s flag, which defines the amount of time that Envoy will
+// drain connections during a hot restart.
+func DrainDuration(duration time.Duration) Option {
+	return &genericOption{
+		value: strconv.Itoa(int(duration / time.Second)),
+		v:     drainDurationValidator,
+	}
+}
+
+var drainDurationValidator = registerFlagValidator(&flagValidator{
+	flagName: "--drain-time-s",
+	validate: func(ctx *configContext, flagValue string) error {
+		if _, err := strconv.ParseUint(flagValue, 10, 32); err != nil {
+			return err
+		}
+		return nil
+	},
+})
+
+// ParentShutdownDuration sets the --parent-shutdown-time-s flag, which defines the amount of
+// time that Envoy will wait before shutting down the parent process during a hot restart
+func ParentShutdownDuration(duration time.Duration) Option {
+	return &genericOption{
+		value: strconv.Itoa(int(duration / time.Second)),
+		v:     parentShutdownDurationValidator,
+	}
+}
+
+var parentShutdownDurationValidator = registerFlagValidator(&flagValidator{
+	flagName: "--parent-shutdown-time-s",
+	validate: func(ctx *configContext, flagValue string) error {
+		if _, err := strconv.ParseUint(flagValue, 10, 32); err != nil {
+			return err
+		}
+		return nil
+	},
+})
+
+// AllowUnknownFields sets the --allow-unknown-fields flag, which disables validation of
+// protobuf configurations for unknown fields
+func AllowUnknownFields(allow bool) Option {
+	var v *flagValidator
+	if allow {
+		v = allowUnknownFieldsValidator
+	}
+
+	return &genericOption{
+		v:     v,
+		value: "",
+	}
+}
+
+var allowUnknownFieldsValidator = registerBoolFlagValidator("--allow-unknown-fields")
+
+func registerBoolFlagValidator(flagName string) *flagValidator {
+	return registerFlagValidator(&flagValidator{
+		flagName: FlagName(flagName),
+		validate: func(ctx *configContext, flagValue string) error {
+			switch flagValue {
+			case "", "true":
+				return nil
+			default:
+				return fmt.Errorf("unexpected boolean value for flag %s: %s", flagName, flagValue)
+			}
+		},
+	})
+}
+
+func getAdminPortFromYaml(yamlData string) (uint32, error) {
+	// TODO(nmittler): The conversion from YAML->JSON doesn't seem to work reliably for the Bootstrap config.
+	jsonData, err := yaml.YAMLToJSON([]byte(yamlData))
```

Comment:
Can we marshal into the bootstrap config struct instead of doing all the `map[string]interface{}` stuff?

Reply:
I tried that initially, but it seems that the conversion from yaml->json either isn't quite correct or that some of you bootstrap yamls have slightly incorrect syntax which yields incorrect json after conversion.  Regardless, we don't seem to be able to correctly parse an envoy bootstrap yaml file using jsonpb.

Reply:
https://github.com/istio/istio/blob/9c9a726e722ab3ad7d544fc4cda47d01e8e18de9/mixer/test/client/gateway/gateway_test.go#L66-L68 should be a list. I think its better to fix there than hack around it here

Reply:
Already switched over to parsing the bootstrap. The mixer test is manually specifying the admin port, so it doesn't impact this PR.